### PR TITLE
feat: Add variant accessor properties to Python bindings

### LINKED ISSUE #45: Add variant accessor properties to Python bindings
# Add variant accessor properties to Python bindings

## Summary

The Python `HgvsVariant` class exposes `variant_type`, `reference`, and `edit_type`, but not the positional or edit-detail data that callers need for common operations (e.g. extracting start/end coordinates, ref/alt alleles, or computing indel length). This data exists in the Rust AST and is available via `to_json()`, but consuming raw serde JSON from Python is fragile and untyped.

Add first-class Python properties that expose the most commonly needed fields directly.

## Motivation

A downstream project currently uses mutalyzer's `to_model()` to parse HGVS strings into structured models. The four primary use cases are:

1. **Positional data** — `model.start` / `model.end` to index into coverage arrays
2. **Edit classification** — checking substitution ref/alt bases to identify common sequencing errors
3. **Allele decomposition** — iterating sub-variants in `[100A>G;200C>T]` alleles
4. **Indel semantics** — computing `indel_length` and `is_frameshift` for frameshift annotation

ferro-hgvs already parses all of this correctly (verified empirically via `to_json()`), but none of it is accessible as Python properties.

## Proposed API additions

Add to `#[pymethods] impl PyHgvsVariant` in `src/python.rs`:

### Positional accessors

```python
@property
def start(self) -> int | None:
    """1-based start position. For alleles, min across sub-variants.
    Returns None for null/unknown allele markers."""

@property
def end(self) -> int | None:
    """1-based end position (inclusive). For alleles, max across sub-variants.
    Returns None for null/unknown allele markers."""

@property
def offset(self) -> int | None:
    """Intronic offset for CDS start position (e.g. 1 for c.93+1G>T).
    Returns None for genomic variants or positions without offsets."""
```

### Edit detail accessors

```python
@property
def substitution_bases(self) -> tuple[str, str] | None:
    """(reference, alternative) for substitutions. None for other edit types."""

def is_identity(self) -> bool:
    """True if the variant is a no-change identity (=) variant."""
```

### Allele decomposition

```python
@property
def num_variants(self) -> int:
    """Number of sub-variants. 1 for simple variants, N for alleles."""

def variants(self) -> list[HgvsVariant]:
    """List of constituent variants. For non-alleles, returns [self]."""
```

### Indel semantics

```python
@property
def indel_length(self) -> int:
    """Net bases inserted (+) or deleted (-). 0 for substitutions/inversions.
    For alleles, sum across sub-variants.
    Computed from interval span and inserted sequence length (no reference lookup needed)."""

def is_frameshift(self) -> bool:
    """True if indel_length % 3 != 0."""
```

### Enrich `to_dict()`

Include the new fields in the dictionary representation:

```python
def to_dict(self) -> dict:
    """Returns: {
        'string': 'NG_008617.1:g.100_102delinsATTT',
        'variant_type': 'genomic',
        'reference': 'NG_008617.1',
        'edit_type': 'delins',
        'start': 100,
        'end': 102,
        'indel_length': 1,
    }"""
```

## Implementation notes

- For `start`/`end` on genomic variants: extract from `GenomeInterval.start.inner().base` / `end.inner().base`
- For `start`/`end` on CDS variants: use `CdsPos.base` (not base+offset — keep them separate)
- For `indel_length`:
  - Substitution/Inversion/Identity → 0
  - Deletion → `-(end - start + 1)`
  - Insertion → length of `InsertedSequence::Literal` (return 0 if sequence is non-literal, e.g. `insN[?]`)
  - Delins → `inserted_len - (end - start + 1)`
  - Duplication → `end - start + 1`
- For alleles: delegate to sub-variants and aggregate (min start, max end, sum indel_length)
- `substitution_bases` → match on `NaEdit::Substitution { reference, alternative }`, return `(ref.to_char(), alt.to_char())`

## Type stub updates

Update `python/ferro_hgvs/__init__.pyi` to include all new properties/methods with full docstrings.

## Testing

- Unit tests in `src/python.rs` (or integration tests) covering:
  - Genomic sub/del/ins/delins/dup/inv/identity
  - CDS variants with and without offsets
  - Allele variants (cis)
  - Edge cases: uncertain positions (`?`), complex inserted sequences
- Python-level tests in `tests/python/` verifying round-trip consistency

### PR DESCRIPTION
**Pull request by msto**
# PR: Add variant accessor properties to Python bindings

**Closes #45**

## Summary

Adds first-class Python properties for extracting positional and edit data from `HgvsVariant` without JSON parsing. These are the fields commonly needed by downstream consumers that previously required mutalyzer's `to_model()` in a downstream project.

### Properties

| Property | Type | Description |
|----------|------|-------------|
| `start` | `int \| None` | 1-based start position (base only, no offset) |
| `end` | `int \| None` | 1-based end position (inclusive) |
| `offset` | `int \| None` | Intronic offset for CDS variants (e.g., 1 for `c.93+1G>T`) |
| `substitution_bases` | `tuple[str, str] \| None` | `("A", "G")` for substitutions |
| `num_variants` | `int` | 1 for simple variants, N for alleles |
| `indel_length` | `int \| None` | Net bases gained/lost (negative for deletions) |

### Methods

| Method | Return | Description |
|--------|--------|-------------|
| `variants()` | `list[HgvsVariant]` | Sub-variants for alleles; `[self]` for simple variants |
| `is_identity()` | `bool` | True for `=` edits |
| `is_frameshift()` | `bool` | True if `indel_length % 3 != 0` |

### Enriched `to_dict()`

```python
>>> ferro_hgvs.parse("NC_000001.11:g.12345A>G").to_dict()
{
    'string': 'NC_000001.11:g.12345A>G',
    'variant_type': 'genomic',
    'reference': 'NC_000001.11',
    'edit_type': 'substitution',
    'start': 12345,
    'end': 12345,
    'ref_base': 'A',
    'alt_base': 'G',
    'indel_length': 0,
    'num_variants': 1
}
```

## Usage examples

```python
import ferro_hgvs

# Position extraction
v = ferro_hgvs.parse("NC_000001.11:g.12345_12350del")
v.start  # 12345
v.end    # 12350
v.indel_length  # -6
v.is_frameshift()  # False (6 % 3 == 0)

# CDS with intronic offset
v = ferro_hgvs.parse("NM_000088.3:c.93+1G>T")
v.start   # 93
v.offset  # 1
v.substitution_bases  # ("G", "T")

# Allele decomposition
v = ferro_hgvs.parse("NM_000088.3:c.[100A>G;200C>T]")
v.num_variants  # 2
for sub in v.variants():
    print(sub.start)  # 100, 200
```

## Implementation

The bulk of the logic lives in `src/python_helpers.rs` as pure Rust functions (testable without the Python runtime), then wired into PyO3 properties in `src/python.rs`.

Key design decisions:
- **`start`/`end` return base only** — offset is a separate property, keeping positional and intronic data orthogonal
- **`indel_length` is computed from the AST** — no reference lookup required. Returns `None` for unknowable cases (uncertain inserted sequences)
- **Allele `start`/`end` delegate to first sub-variant** — simple and matches the common case where all sub-variants share a reference
- **`to_dict()` is additive** — all old keys remain; new keys appear only when applicable (e.g., `ref_base`/`alt_base` only for substitutions)

## Testing

- 26 new unit tests in `python_helpers.rs` covering all accessors, edge cases, and indel length arithmetic
- All 1875 lib tests pass
- All 80 existing Python tests pass unchanged
- Type stubs updated in `__init__.pyi`

## Files changed

- `src/python_helpers.rs` — new helper functions (position extraction, edit accessors, indel length)
- `src/python.rs` — new `#[getter]` properties and methods on `PyHgvsVariant`
- `python/ferro_hgvs/__init__.pyi` — type stub updates

### CODE REVIEW

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

File: src/python_helpers.rs [hunk 1]
```diff
@@ -199,6 +201,210 @@ pub fn get_variant_edit_type(variant: &HgvsVariant) -> &'static str {
...
+    }
+}
+
+/// Get the intronic offset of the start position for CDS variants.
+///
+/// For c.93+1G>T, returns Some(1). For c.100A>G (exonic), returns None.
+/// Only meaningful for CDS variants; returns None for all other types.
+pub fn get_variant_offset(variant: &HgvsVariant) -> Option<i64> {
+    match variant {
+        HgvsVariant::Cds(v) => v.loc_edit.location.start.inner().and_then(|pos| pos.offset),
```
Comment (nh13) [inline src/python_helpers.rs:309]: **issue:** add `TxPos` and `RnaPos` too, they have offsets and would feel sad to be left out

File: src/python_helpers.rs [hunk 2]
```diff
@@ -199,6 +201,210 @@ pub fn get_variant_edit_type(variant: &HgvsVariant) -> &'static str {
...
+fn rna_interval_end(interval: &Interval<RnaPos>) -> Option<i64> {
+    interval.end.inner().map(|pos| pos.base)
+}
+
+/// Get the start position (base only, no offset) for any variant.
+///
+/// Returns the 1-based start position for genomic, coding, non-coding, RNA, and
+/// mitochondrial variants. For alleles, returns the start of the first sub-variant.
+/// Returns None for protein, fusion, null/unknown allele variants.
+pub fn get_variant_start(variant: &HgvsVariant) -> Option<i64> {
```
Comment (nh13) [inline src/python_helpers.rs:257]: **nitpick:** `c.*5A>G` and `c.5A>G` are indistinguishable, worth a docstring comment?

File: src/python_helpers.rs [hunk 3]
```diff
@@ -199,6 +201,210 @@ pub fn get_variant_edit_type(variant: &HgvsVariant) -> &'static str {
...
+
+/// Extract the end base position from an RNA interval
+fn rna_interval_end(interval: &Interval<RnaPos>) -> Option<i64> {
+    interval.end.inner().map(|pos| pos.base)
+}
+
+/// Get the start position (base only, no offset) for any variant.
+///
+/// Returns the 1-based start position for genomic, coding, non-coding, RNA, and
+/// mitochondrial variants. For alleles, returns the start of the first sub-variant.
```
Comment (nh13) [inline src/python_helpers.rs]: **nitpick:** consider returning `None` if multiple sub-variants exist, or a `Vec<i64>`?

File: src/python_helpers.rs [hunk 4]
```diff
@@ -199,6 +201,210 @@ pub fn get_variant_edit_type(variant: &HgvsVariant) -> &'static str {
...
+
+/// Check if the variant represents an identity (no-change, `=`) edit.
+pub fn is_identity(variant: &HgvsVariant) -> bool {
+    matches!(get_na_edit(variant), Some(NaEdit::Identity { .. }))
+}
+
+/// Compute the span of the affected region from the interval.
+/// For point variants (start == end), the span is 1 for deletion/delins
+/// but 0 for insertions (which occur between two positions).
+fn compute_span(variant: &HgvsVariant) -> Option<i64> {
```
Comment (nh13) [inline src/python_helpers.rs:358]: **issue:** `HgvsVariant::Circular` will return the wrong span.  Consider adding a unit test and returning `None` when `end < start`

File: src/python_helpers.rs [hunk 5]
```diff
@@ -199,6 +201,210 @@ pub fn get_variant_edit_type(variant: &HgvsVariant) -> &'static str {
...
+}
+
+/// Check if the variant represents an identity (no-change, `=`) edit.
+pub fn is_identity(variant: &HgvsVariant) -> bool {
+    matches!(get_na_edit(variant), Some(NaEdit::Identity { .. }))
+}
+
+/// Compute the span of the affected region from the interval.
+/// For point variants (start == end), the span is 1 for deletion/delins
+/// but 0 for insertions (which occur between two positions).
```
Comment (nh13) [inline src/python_helpers.rs]: **question:** is `compute_span` used for insertions?

File: src/python.rs [hunk 6]
```diff
@@ -132,6 +134,100 @@ impl PyHgvsVariant {
...
+    fn offset(&self) -> Option<i64> {
+        get_variant_offset(&self.inner)
+    }
+
+    /// Get the substitution reference and alternative bases.
+    ///
+    /// Returns a tuple (ref_base, alt_base) for substitution edits,
+    /// e.g., ("A", "G") for A>G. Returns None for non-substitution edits.
+    #[getter]
+    fn substitution_bases(&self) -> Option<(String, String)> {
```
Comment (nh13) [inline src/python.rs]: **nitpick:** if truly single base, why not:
```suggestion
    fn substitution_bases(&self) -> Option<(char, char)> {
```

File: src/python_helpers.rs [hunk 7]
```diff
@@ -513,4 +719,178 @@ mod tests {
         let variant = HgvsVariant::UnknownAllele;
         assert_eq!(get_variant_edit_type(&variant), "unknown");
     }
+
+    // ===== Position Accessor Tests =====
```
Comment (nh13) [inline src/python_helpers.rs:733]: **issue:** 

Consider adding tests:

- No n. (Tx) / r. (Rna) tests: existing offset tests only at :757-772 (CDS).
- No UTR test (c.*5A>G) for start/end
- No protein-variant assertions for start/end/indel_length returning None.
- No Insertion test with non-literal sequence (e.g., Count/Range returning None) — inserted_sequence_len (:386-388) is only exercised via the Literal
- path at :830-833 and :836-840.
- No Python-level tests in tests/python/ covering the new properties or to_dict() keys (src/python.rs:321-341).
- Allele test only at src/python_helpers.rs:775-779: single accession, doesn't probe sub-variant divergence.

**Review by nh13 — CHANGES_REQUESTED** [review]

File: src/python_helpers.rs (hunk 5 above)
Comment (msto) [reply src/python_helpers.rs]: No — insertions get their length from the inserted-sequence AST via `inserted_sequence_len`, not from start/end. Only deletion, duplication, and delins call `compute_span`. I updated the function's docstring to spell that out.

**Review by nh13 — APPROVED** [review]
